### PR TITLE
[Fix] maintain connections; graceful shutdown

### PR DESCRIPTION
**Pull request by bojiang**
## Description

Closing #1618 
Fix for the marshal server:
* Maintain outbound connections
* Release coroutine manually to shutdown outbound connections gracefully (or we would get an exception on exit)

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature and improvements (non-breaking change which adds/improves functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Code Refactoring (internal change which is not user facing)
- [ ] Documentation
- [ ] Test, CI, or build

## Component(s) if applicable
- [x] BentoService (service definition, dependency management, API input/output adapters)
- [ ] Model Artifact (model serialization, multi-framework support)
- [ ] Model Server (mico-batching, dockerisation, logging, OpenAPI, instruments)
- [ ] YataiService gRPC server (model registry, cloud deployment automation)
- [ ] YataiService web server (nodejs HTTP server and web UI)
- [ ] Internal (BentoML's own configuration, logging, utility, exception handling)
- [ ] BentoML CLI

## Checklist:

- [x] My code follows the bentoml code style, both `./dev/format.sh` and
  `./dev/lint.sh` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [ ] My change reduces project test coverage and requires unit tests to be added
- [ ] I have added unit tests covering my code change
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: bentoml/marshal/marshal.py [hunk 1]
```diff
@@ -213,23 +216,36 @@ def fetch_sema(self):
...
         return self._client
 
     def __del__(self):
+        loop = asyncio.get_event_loop()
         if getattr(self, '_client', None) is not None and not self._client.closed:
-            self._client.close()
+            loop.create_task(self._client.close())
+
+        # https://docs.aiohttp.org/en/stable/client_advanced.html#graceful-shutdown
+        loop.create_task(asyncio.sleep(0))
```
Comment (parano) [inline bentoml/marshal/marshal.py:248]: what's the difference between this and what was suggested in the docs: `loop.run_until_complete(asyncio.sleep(0))`?

Comment (bojiang) [reply bentoml/marshal/marshal.py:248]: When the loop is already running, we append a new task to it rather than `loop.run_until_complete`

**Comment by aarnphm** [comment]
Is #1618 related to this @bojiang?

**Comment by aarnphm** [comment]
can confirmed that #1618 is related. When closing returned `RuntimeError`:

```bash

^C[2021-06-13 05:46:06 +0000] [1] [INFO] Handling signal: int                                                                                                                                                                                                                               [2021-06-13 05:46:06 +0000] [22] [INFO] Worker exiting (pid: 22)
[2021-06-13 05:46:06 +0000] [20] [INFO] Handling signal: term
[2021-06-13 05:46:06 +0000] [21] [INFO] Worker exiting (pid: 21)
[2021-06-13 05:46:06 +0000] [20] [INFO] Shutting down: Master
Exception ignored in: <function MarshalService.__del__ at 0x7f250c6b9430>
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/bentoml/marshal/marshal.py", line 248, in __del__
    loop.create_task(asyncio.sleep(0))
  File "/opt/conda/lib/python3.8/asyncio/base_events.py", line 429, in create_task
    self._check_closed()
  File "/opt/conda/lib/python3.8/asyncio/base_events.py", line 508, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
[2021-06-13 05:46:07 +0000] [1] [INFO] Shutting down: Master
```

**Comment by aarnphm** [comment]
can confirmed that #1618 is related. When closing returned `RuntimeError`:

```bash

^C[2021-06-13 05:46:06 +0000] [1] [INFO] Handling signal: int                                                                                                                                                                                                                               [2021-06-13 05:46:06 +0000] [22] [INFO] Worker exiting (pid: 22)
[2021-06-13 05:46:06 +0000] [20] [INFO] Handling signal: term
[2021-06-13 05:46:06 +0000] [21] [INFO] Worker exiting (pid: 21)
[2021-06-13 05:46:06 +0000] [20] [INFO] Shutting down: Master
Exception ignored in: <function MarshalService.__del__ at 0x7f250c6b9430>
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/site-packages/bentoml/marshal/marshal.py", line 248, in __del__
    loop.create_task(asyncio.sleep(0))
  File "/opt/conda/lib/python3.8/asyncio/base_events.py", line 429, in create_task
    self._check_closed()
  File "/opt/conda/lib/python3.8/asyncio/base_events.py", line 508, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
[2021-06-13 05:46:07 +0000] [1] [INFO] Shutting down: Master
```

**Comment by parano** [comment]
@bojiang could you take a look when you go a chance? I'm seeing the same error message that @aarnphm had with this branch

**Comment by bojiang** [comment]
Yeah. I see. This error makes sense for me. I'll check it again.

**Comment by bojiang** [comment]
@parano ready for review

File: bentoml/marshal/marshal.py (hunk 1 above)
Comment (ssheng) [reply bentoml/marshal/marshal.py]: > Note that the appropriate amount of time to wait will vary from application to application.

It seems like no amount of time is reliable. Can we make the sleep time configurable and default to something conservative? Also, `aiohttp` 4.0 has resolved this issue. Let's keep it in the comments so when we upgrade we will know to fix here as well.

File: bentoml/marshal/dispatcher.py [hunk 2]
```diff
@@ -120,6 +120,17 @@ def __init__(
         self._queue = collections.deque()  # TODO(hrmthw): maxlen
         self._sema = shared_sema if shared_sema else NonBlockSema(1)
 
+    async def cleanup(self):
+        if self._controller is not None:
+            self._controller.cancel()
+            await asyncio.sleep(0)
+        try:
+            while True:
+                _, _, fut = self._queue.pop()
```
Comment (ssheng) [inline bentoml/marshal/dispatcher.py:129]: Should we also add a condition to also reject all inbound calls?

File: bentoml/marshal/dispatcher.py [hunk 3]
```diff
@@ -120,6 +120,17 @@ def __init__(
         self._queue = collections.deque()  # TODO(hrmthw): maxlen
         self._sema = shared_sema if shared_sema else NonBlockSema(1)
 
+    async def cleanup(self):
```
Comment (ssheng) [inline bentoml/marshal/dispatcher.py:123]: nit: `shutdown()` may be more accurate here than `cleanup()`

File: bentoml/marshal/dispatcher.py [hunk 4]
```diff
@@ -120,6 +120,17 @@ def __init__(
...
+        if self._controller is not None:
+            self._controller.cancel()
+            await asyncio.sleep(0)
```
Comment (ssheng) [inline bentoml/marshal/dispatcher.py:126]: Is `asyncio.sleep(0)` needed for connection close?

File: bentoml/marshal/dispatcher.py (hunk 2 above)
Comment (bojiang) [reply bentoml/marshal/dispatcher.py:129]: There will not be race conditions in an async APP.

File: bentoml/marshal/dispatcher.py (hunk 4 above)
Comment (bojiang) [reply bentoml/marshal/dispatcher.py:126]: Yep, it's unnecessary here.

File: bentoml/marshal/marshal.py (hunk 1 above)
Comment (bojiang) [reply bentoml/marshal/marshal.py]: @ssheng Ends up by hooking on the aiohttp `on_cleanup` to solve this.

**Pull request merged**